src/boot/libu64/gfxprint.c

```diff
@@ -0,0 +1,262 @@
+    }
+
+    gDPSetColor(this->gListp++, G_SETPRIMCOLOR, this->color.rgba);
+
+    gDPLoadMultiTile_4b(this->gListp++, gfxprint_rainbow_txtr, 0, 1, G_IM_FMT_CI, 2, 8, 0, 0, 1, 7, 4,
+                        G_TX_NOMIRROR | G_TX_WRAP, G_TX_NOMIRROR | G_TX_WRAP, 1, 3, G_TX_NOLOD, G_TX_NOLOD);
+
+    // loads 16 texels to palette index 4
+    gDPLoadTLUT(this->gListp++, 16, 0x100 + 4 * 0x10, gfxprint_rainbow_tlut);
+
+    for (i = 1; i < 4; i++) {
+        gDPSetTile(this->gListp++, G_IM_FMT_CI, G_IM_SIZ_4b, TX_LINE(16, G_IM_SIZ_4b), 0, i * 2 + 1, 4,
+                   G_TX_NOMIRROR | G_TX_WRAP, 3, G_TX_NOLOD, G_TX_NOMIRROR | G_TX_WRAP, 1, G_TX_NOLOD);
+        gDPSetTileSize(this->gListp++, i * 2 + 1, 0, 0, (2 - 1) << G_TEXTURE_IMAGE_FRAC,
+                       (8 - 1) << G_TEXTURE_IMAGE_FRAC);
+    }
+}
+
+void gfxprint_color(gfxprint* this, u32 r, u32 g, u32 b, u32 a) {
+    this->color.r = r;
+    this->color.g = g;
+    this->color.b = b;
+    this->color.a = a;
+
+    gDPPipeSync(this->gListp++);
+    gDPSetColor(this->gListp++, G_SETPRIMCOLOR, this->color.rgba);
+}
+
+void gfxprint_locate(gfxprint* this, s32 x, s32 y) {
+    this->positionX = this->offsetX + x * 4;
+    this->positionY = this->offsetY + y * 4;
+}
+
+void gfxprint_locate8x8(gfxprint* this, s32 x, s32 y) {
+    gfxprint_locate(this, x * 8, y * 8);
+}
+
+void gfxprint_setoffset(gfxprint* this, s32 x, s32 y) {
+    this->offsetX = x * 4;
+    this->offsetY = y * 4;
+}
+
+void gfxprint_putc1(gfxprint* this, char c) {
+    u32 tile = (c & 0xFF) * 2;
+    u16 x0 = (c & 4);
+    u16 x1 = (c >> 3);
+
+    if (gfxprint_isChanged(this)) {
+        gfxprint_clrChanged(this);
+        gDPPipeSync(this->gListp++);
+
+        if (gfxprint_isGradient(this)) {
+            gDPSetTextureLUT(this->gListp++, G_TT_RGBA16);
+            gDPSetCycleType(this->gListp++, G_CYC_2CYCLE);
+            gDPSetRenderMode(this->gListp++, G_RM_PASS, G_RM_XLU_SURF2);
+            gDPSetCombineMode(this->gListp++, G_CC_INTERFERENCE, G_CC_PASS2);
+        } else {
+            gDPSetTextureLUT(this->gListp++, G_TT_IA16);
+            gDPSetCycleType(this->gListp++, G_CYC_1CYCLE);
+            gDPSetRenderMode(this->gListp++, G_RM_XLU_SURF, G_RM_XLU_SURF2);
+            gDPSetCombineMode(this->gListp++, G_CC_MODULATEIDECALA_PRIM, G_CC_MODULATEIDECALA_PRIM);
+        }
+    }
+
+    if (gfxprint_isShadow(this)) {
+
+        gDPSetColor(this->gListp++, G_SETPRIMCOLOR, 0);
+
+        if (gfxprint_isHighres(this)) {
+            gSPTextureRectangle(this->gListp++, (this->positionX + 4) << 1, (this->positionY + 4) << 1,
+                                (this->positionX + 4 + 32) << 1, (this->positionY + 4 + 32) << 1, tile, x0 << 6,
+                                x1 << 8, 512, 512);
+        } else {
+            gSPTextureRectangle(this->gListp++, this->positionX + 4, this->positionY + 4, this->positionX + 4 + 32,
+                                this->positionY + 4 + 32, tile, x0 << 6, x1 << 8, 1024, 1024);
```

Comment:
```suggestion
            gSPTextureRectangle(this->gListp++, (this->positionX + 4) << 1, (this->positionY + 4) << 1,
                                (this->positionX + 4 + 32) << 1, (this->positionY + 4 + 32) << 1, tile, x0 << 6,
                                x1 << 8, 1 << 9, 1 << 9);
        } else {
            gSPTextureRectangle(this->gListp++, this->positionX + 4, this->positionY + 4, this->positionX + 4 + 32,
                                this->positionY + 4 + 32, tile, x0 << 6, x1 << 8, 1 << 10, 1 << 10);
```
Same for the two other `gSPTextureRectangle`s below please :)